k2/csrc/hash.h

```diff
@@ -1014,6 +1014,349 @@ class Hash {
+      reasonable hash occupancy levels, we shouldn't ever have to try
+      more than two or three).
+
+*/
+class Hash64 {
+ public:
+  /* Constructor.  Context can be for CPU or GPU.
+
+     @param [in] num_buckets   Number of buckets in the hash; must be
+                a power of 2 and >= 128 (this limit was arbitrarily chosen).
+                The number of items in the hash cannot exceed the number of
+                buckets, or the code will loop infinitely when you try to add
+                items; aim for less than 50% occupancy.
+  */
+  Hash64(ContextPtr c, int32_t num_buckets) {
+    K2_CHECK_GE(num_buckets, 128);
+    data_ = Array1<uint64_t>(c, num_buckets * 2, ~(uint64_t)0);
+    int32_t n = 2;
+    for (buckets_num_bitsm1_ = 0; n < num_buckets;
+         n *= 2, buckets_num_bitsm1_++) {
+    }
+    K2_CHECK_EQ(num_buckets, 2 << buckets_num_bitsm1_)
+        << " num_buckets must be a power of 2.";
+  }
+
+  // Only to be used prior to assignment.
+  Hash64() = default;
+
+  int32_t NumBuckets() const { return data_.Dim() / 2; }
+
+  // Returns data pointer; for testing..
+  uint64_t *Data() { return data_.Data(); }
+
+  // Shallow copy
+  Hash64 &operator=(const Hash64 &src) = default;
+  // Copy constructor (shallow copy)
+  explicit Hash64(const Hash64 &src) = default;
+
+  ContextPtr &Context() const { return data_.Context(); }
+
+  class Accessor {
+   public:
+    Accessor(Hash64 &hash)
+        : data_(hash.data_.Data()),
+          num_buckets_mask_(uint32_t(hash.NumBuckets()) - 1),
+          buckets_num_bitsm1_(hash.buckets_num_bitsm1_) {}
+
+    // Copy constructor
+    Accessor(const Accessor &src) = default;
+
+    /*
+     Try to insert pair (key,value) into hash.
+       @param [in] key  Key into hash, it is an error if ~key == 0, i.e. if all
+                        the allowed bits of `key` are set.
+       @param [in] value  Value to set, it is an error if ~value == 0, i.e. if
+                          all the allowed bits `value` are set.
+       @param [out] old_value  If not nullptr, this location will be set to
+                     the existing value *if this key was already present* in the
+                     hash (or set by another thread in this kernel), i.e. only
+                     if this function returns false.
+       @param [out] key_value_location  If not nullptr, its contents will be
+                     set to the address of the (key,value) pair (either the
+                     existing or newly-written one).
+       @return  Returns true if this (key,value) pair was inserted, false
+     otherwise.
+
+       Note: the const is with respect to the metadata only; it is required, to
+       avoid compilation errors.
+    */
+    __forceinline__ __host__ __device__ bool Insert(
+        uint64_t key, uint64_t value, uint64_t *old_value = nullptr,
+        uint64_t **key_value_location = nullptr) const {
+      uint32_t cur_bucket = static_cast<uint32_t>(key) & num_buckets_mask_,
```

Comment:
Any reason to discard the higher 32bits of the key?
There is no such a cast at line 1178.

Reply:
Maybe, it is better to change the bucket index to unint64_t.

Reply:
I think num_buckets_mask_ is uint32_t. 
I think we cannot index larger than int32_t into Array, so while it would be possible to make both those things uint64_t, I'm not sure that it would have a benefit.  But it might reduce the number of instructions slightly if we made them both uint64_t, so I  guess I'd be OK with that??  Or at least cast num_buckets_mask_ to uint64_t before using it...